src/lib/transaction_snark/transaction_snark.ml

```diff
@@ -2309,24 +2310,29 @@ module Base = struct
                     add fee_payer_amount account_creation_fee))
              in
              let txn_global_slot = current_global_slot in
-             let%bind `Min_balance _, timing =
-               [%with_label "Check fee payer timing"]
-                 (let%bind txn_amount =
-                    Amount.Checked.if_
-                      (Sgn.Checked.is_neg amount.sgn)
-                      ~then_:amount.magnitude
-                      ~else_:Amount.(var_of_t zero)
-                  in
-                  let balance_check ok =
-                    [%with_label "Check fee payer balance"]
-                      (Boolean.Assert.is_true ok)
-                  in
-                  let timed_balance_check ok =
-                    [%with_label "Check fee payer timed balance"]
-                      (Boolean.Assert.is_true ok)
-                  in
-                  check_timing ~balance_check ~timed_balance_check ~account
-                    ~txn_amount ~txn_global_slot)
+             let%bind timing =
+               let%bind `Min_balance _, new_timing =
+                 [%with_label "Check fee payer timing"]
+                   (let%bind txn_amount =
+                      Amount.Checked.if_
+                        (Sgn.Checked.is_neg amount.sgn)
+                        ~then_:amount.magnitude
+                        ~else_:Amount.(var_of_t zero)
+                    in
+                    let balance_check ok =
+                      [%with_label "Check fee payer balance"]
+                        (Boolean.Assert.is_true ok)
+                    in
+                    let timed_balance_check ok =
+                      [%with_label "Check fee payer timed balance"]
+                        (Boolean.Assert.is_true ok)
+                    in
+                    check_timing ~balance_check ~timed_balance_check ~account
+                      ~txn_amount ~txn_global_slot)
+               in
+               (* Like receiver account in payments, coinbase fee transfer account or other fee transfer account don't have the timing field updated *)
+               Account_timing.if_ is_coinbase_or_fee_transfer
```

Comment:
This `if_` call is rather expensive. Can we rework this to change the non-snark logic instead?

Reply:
If we update both accounts in coinbase and fee transfers (otherwise it seems inconsistent) then there will still be an `if_`  call when we update the receiver timing (if user command then don't update the timing). An alternative would be to just update all the accounts' timing field for every type of transaction (stake delegation, receiver account in payments, coinbase, and fee transfer)

Reply:
> An alternative would be to just update all the accounts' timing field for every type of transaction (stake delegation, receiver account in payments, coinbase, and fee transfer)

This is definitely my preferred solution. This should already be the case for all user commands, it should just be coinbases and fee transfers that need updating in `transaction_logic.ml` and `sparse_ledger.ml`.

Reply:
For user commands, we currently don't update the receiver account's timing (receiver in payments and delegate in stake delegations). i'll change that as well

Reply:
@mrmr1993 Would checking the timing for receiver account be cheaper than the two `if_` calls though?

Reply:
Do we need the receiver account timing to be updated? I don't think we ever reduce the balance for the receiver account, so I would have thought the timing isn't relevant, unless I'm missing some nuance.

Reply:
Well, we don't reduce the balance for coinbase and fee transfer accounts either. But the way we encode those as `fee_payer` and `receiver`, we either update both or none otherwise it is inconsistent. Like, one of the fee transfers in a fee transfer transaction has the timing updated but not the other

Reply:
In those cases, I had leant on the side of doing what's cheapest in the snark. It seems correct to me to either update the timing or not when we're not decreasing the balance of an account. Since we always re-check the timing whenever it's relevant, it should make no practical difference.

I propose changing the non-snark logic to match the existing snark behaviour, and that we live with the small inconsistency around the receiver account not getting it's timing re-normalised.

Reply:
Alright, makes sense in the interest of not increasing the proving time whenever possible